rest_log/tests/test_db_logging.py

```diff
@@ -204,3 +228,102 @@ def test_log_entry_severity_mapping_param_bad_values(self):
+                self.assertTrue(isinstance(err, wrapping_exc))
+                self.assertEqual(
+                    self.service._get_exception_message(err), "Failed as you wanted!"
+                )
+                entry_url_from_exc = err.rest_json_info["log_entry_url"]
+
+        with new_rollbacked_env() as env:
+            log_model = env["rest.log"].sudo()
+            entry = log_model.search([]) - initial_entries
+            expected = {
+                "collection": self.service._collection,
+                "state": "failed",
+                "result": "null",
+                "exception_name": exc_name,
+                "exception_message": "Failed as you wanted!",
+                "severity": severity,
+            }
+            self.assertRecordValues(entry, [expected])
+            self.assertEqual(entry_url_from_exc, self.service._get_log_entry_url(entry))
+
+
+class TestDBLoggingExceptionUserError(TestDBLoggingExceptionBase):
+    @staticmethod
+    def _get_test_controller(class_or_instance, root_path=None):
+        # Override to avoid registering twice the same controller route.
+        # Disclaimer: to run these tests w/ need TransactionCase
+        # because the handling of the exception will do a savepoint rollback
+        # which causes SavepointCase to fail.
+        # When using the transaction case the rest_registry is initliazed
+        # at every test, same for the test controller.
+        # This leads to the error
+        # "Only one REST controller
+        # can be safely declared for root path /test_controller/"
+        return super()._get_test_controller(
+            class_or_instance, root_path="/test_log_exception_user/"
+        )
+
+    def test_log_exception_user(self):
+        self._test_exception(
+            "user",
+            log_exceptions.RESTServiceUserErrorException,
+            "odoo.exceptions.UserError",
+            "functional",
+        )
+
+
+class TestDBLoggingExceptionValidationError(TestDBLoggingExceptionBase):
+    @staticmethod
+    def _get_test_controller(class_or_instance, root_path=None):
+        return super()._get_test_controller(
+            class_or_instance, root_path="/test_log_exception_validation/"
+        )
+
+    def test_log_exception_validation(self):
+        self._test_exception(
+            "validation",
+            log_exceptions.RESTServiceValidationErrorException,
+            "odoo.exceptions.ValidationError",
+            "functional",
+        )
+
+
+class TestDBLoggingExceptionValueError(TestDBLoggingExceptionBase):
+    @staticmethod
+    def _get_test_controller(class_or_instance, root_path=None):
+        return super()._get_test_controller(
+            class_or_instance, root_path="/test_log_exception_value/"
+        )
+
+    def test_log_exception_value(self):
+        self._test_exception(
+            "value", log_exceptions.RESTServiceDispatchException, "ValueError", "severe"
+        )
```

Comment:
@simahawk these tests are failing and I don't have enough context about error log. Pls help to fix.
```
odoo/src/odoo/sql_db.py:89: in check
    return f(self, *args, **kwargs)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <odoo.sql_db.Cursor object at 0x7f2a68d063a0>, query = 'ROLLBACK TO SAVEPOINT test_10', params = None, log_exceptions = None

    @check
    def execute(self, query, params=None, log_exceptions=None):
        if params and not isinstance(params, (tuple, list, dict)):
            # psycopg2's TypeError is not clear if you mess up the params
            raise ValueError("SQL query parameters should be a tuple, list or dict; got %r" % (params,))
    
        if self.sql_log:
            _logger.debug("query: %s", self._format(query, params))
        start = time.time()
        try:
            params = params or None
>           res = self._obj.execute(query, params)
E           psycopg2.errors.InvalidSavepointSpecification: savepoint "test_10" does not exist

odoo/src/odoo/sql_db.py:310: InvalidSavepointSpecification
```

Reply:
hmm the problem comes from the fact that now `TransactionCase` uses a savepoint,

These tests ensure that when an exception is catched while executing an endpoint, a rest.log entry is created.
In a SavepointCase the savepoint of the test case is rolled back immediately.

We could try using `SingleTransactionCase` which uses a single transaction w/out savepoint.

Option B: we could try using a savepoint [here](https://github.com/OCA/rest-framework/blob/14.0/rest_log/components/service.py#L42) as suggested by @lmignon in the past (see the comment right above).

Instead of using a try/except block we could use a savepoint and rollback to it. This way the original transaction is not invalidated. But: I'm not sure it will solve the problem and it implies a refactoring.